File: src/graph_models/app.model.js
import { neo4jDriver } from "../configs/graph_db.js"

export class App {
  static async searchAndFilterPosts({
    term,
    filter,
    limit,
    offset,
    client_user_id,
  }) {
    let applyFilter = ""

    if (filter !== "all") {
      applyFilter = "post.type = $filter AND "
    }

    const { records } = await neo4jDriver.executeRead(
      `
        MATCH (post:Post WHERE ${applyFilter}post.description CONTAINS $term)<-[:CREATES_POST]-(ownerUser:User), (clientUser:User{ id: $client_user_id })
    
        OPTIONAL MATCH (clientUser)-[crxn:REACTS_TO_POST]->(post)
        OPTIONAL MATCH (clientUser)-[csaves:SAVES_POST]->(post)
        OPTIONAL MATCH (clientUser)-[creposts:REPOSTS_POST]->(post)
    
        WITH post, 
          toString(post.created_at) AS created_at, 
          ownerUser { .id, .username, .profile_pic_url } AS owner_user,
          CASE crxn 
            WHEN IS NULL THEN "" 
            ELSE crxn.reaction 
          END AS client_reaction, 
          CASE csaves 
            WHEN IS NULL THEN false 
            ELSE true 
          END AS client_saved, 
          CASE creposts 
            WHEN IS NULL THEN false 
            ELSE true 
          END AS client_reposted
        ORDER BY post.created_at DESC
        OFFSET $offset
        LIMIT $limit
        RETURN collect(post { .*, owner_user, created_at, client_reaction, client_saved, client_reposted }) AS search_results
        `,
      { term, filter, client_user_id, limit, offset }
    )

    return records[0].get("search_results")
  }

  static async searchHashtags({ term, limit, offset }) {
    const { records } = await neo4jDriver.executeRead(
      `
      MATCH (ht:Hashtag WHERE ht.name CONTAINS $term)<-[:INCLUDES_HASHTAG]-(post:Post)
      WITH ht.name AS hashtag, count(post) AS posts_count
      ORDER BY post_count DESC
      OFFSET $offset
      LIMIT $limit
      RETURN collect({ hashtag, posts_count }) AS search_results
      `,
      { term, limit, offset }
    )

    return records[0].get("search_results")
  }

  static async searchUsers({ term, limit, offset }) {
    const { records } = await neo4jDriver.executeRead(
      `
      MATCH (user:User WHERE user.username CONTAINS $term OR user.name CONTAINS $term)
      ORDER BY user.username, user.name
      OFFSET $offset
      LIMIT $limit
      RETURN collect(user { .username, .name, .profile_pic_url, .bio }) AS search_results
      `,
      { term, limit, offset }
    )

    return records[0].get("search_results")
  }

  static async getHashtagPosts({
    hashtag_name,
    filter,
    limit,
    offset,
    client_user_id,
  }) {
    let applyFilter = ""

    if (filter != "all") {
      applyFilter = " WHERE post.type = $filter"
    }

    const { records } = await neo4jDriver.executeRead(
      `
      MATCH (ht:Hashtag{ name: $hashtag_name })<-[:INCLUDES_HASHTAG]-(post:Post${applyFilter})<-[:CREATES_POST]-(ownerUser:User), (clientUser:User{ id: $client_user_id })

      OPTIONAL MATCH (clientUser)-[crxn:REACTS_TO_POST]->(post)
      OPTIONAL MATCH (clientUser)-[csaves:SAVES_POST]->(post)
      OPTIONAL MATCH (clientUser)-[creposts:REPOSTS_POST]->(post)
  
      WITH post, 
        toString(post.created_at) AS created_at, 
        ownerUser { .id, .username, .profile_pic_url } AS owner_user,
        CASE crxn 
          WHEN IS NULL THEN "" 
          ELSE crxn.reaction 
        END AS client_reaction, 
        CASE csaves 
          WHEN IS NULL THEN false 
          ELSE true 
        END AS client_saved, 
        CASE creposts 
          WHEN IS NULL THEN false 
          ELSE true 
        END AS client_reposted
      ORDER BY post.created_at DESC
      OFFSET $offset
      LIMIT $limit
      RETURN collect(post { .*, owner_user, created_at, client_reaction, client_saved, client_reposted }) AS search_results
      `,
      { filter, hashtag_name, limit, offset, client_user_id }
    )

    return records[0].get("search_results")
  }
}

File: src/models/appModel/appModel.go
package appModel

import (
	"context"
	"i9lyfe/src/models/db"
	"log"

	"github.com/gofiber/fiber/v2"
	"github.com/neo4j/neo4j-go-driver/v5/neo4j"
)

func TopUsers(ctx context.Context) ([]any, error) {
	res, err := db.Query(
		ctx,
		`
    MATCH (user:User)<-[:FOLLOWS_USER]-(follower:User)

		WITH user, count(follower) AS followers_count
    ORDER BY followers_count DESC
    LIMIT 500
    RETURN collect(user { .username, .name, .profile_pic_url, .bio }) AS top_users
    `,
		nil,
	)
	if err != nil {
		log.Println("appModel.go: TopUsers:", err)
		return nil, fiber.ErrInternalServerError
	}

	if len(res.Records) == 0 {
		return nil, nil
	}

	topUsers, _, _ := neo4j.GetRecordValue[[]any](res.Records[0], "top_users")

	return topUsers, nil
}

func SearchUsers(ctx context.Context, term string) ([]any, error) {
	res, err := db.Query(
		ctx,
		`
    MATCH (user:User WHERE user.username CONTAINS $term OR user.name CONTAINS $term)

    ORDER BY user.username, user.name
    LIMIT 500
    RETURN collect(user { .username, .name, .profile_pic_url, .bio }) AS res_users
    `,
		map[string]any{
			"term": term,
		},
	)
	if err != nil {
		log.Println("appModel.go: SearchUsers:", err)
		return nil, fiber.ErrInternalServerError
	}

	if len(res.Records) == 0 {
		return nil, nil
	}

	resUsers, _, _ := neo4j.GetRecordValue[[]any](res.Records[0], "res_users")

	return resUsers, nil
}

func TopHashtags(ctx context.Context) ([]any, error) {
	res, err := db.Query(
		ctx,
		`
    MATCH (ht:Hashtag)<-[:INCLUDES_HASHTAG]-(post:Post)

		WITH ht.name AS hashtag, count(post) AS posts_count
		ORDER BY post_count DESC
		LIMIT 1000
		RETURN collect({ hashtag, posts_count }) AS top_hashtags
    `,
		nil,
	)
	if err != nil {
		log.Println("appModel.go: TopHashtags:", err)
		return nil, fiber.ErrInternalServerError
	}

	if len(res.Records) == 0 {
		return nil, nil
	}

	topUsers, _, _ := neo4j.GetRecordValue[[]any](res.Records[0], "top_hashtags")

	return topUsers, nil
}

func SearchHashtags(ctx context.Context, term string) ([]any, error) {
	res, err := db.Query(
		ctx,
		`
    MATCH (ht:Hashtag WHERE ht.name CONTAINS $term)<-[:INCLUDES_HASHTAG]-(post:Post)

		WITH ht.name AS hashtag, count(post) AS posts_count
		ORDER BY post_count DESC
		LIMIT 1000
		RETURN collect({ hashtag, posts_count }) AS res_hashtags
    `,
		map[string]any{
			"term": term,
		},
	)
	if err != nil {
		log.Println("appModel.go: SearchHashtags:", err)
		return nil, fiber.ErrInternalServerError
	}

	if len(res.Records) == 0 {
		return nil, nil
	}

	topUsers, _, _ := neo4j.GetRecordValue[[]any](res.Records[0], "top_hashtags")

	return topUsers, nil
}

func SearchPost(ctx context.Context, clientUsername, postType, term string) ([]any, error) {
	var query = ""

	if clientUsername != "" {
		query = `
			MATCH (clientUser:User{ username: $client_username })
			MATCH (ownerUser:User)-[:CREATES_POST]->(post:Post { type: $type } WHERE post.description CONTAINS $term)
	
			OPTIONAL MATCH (clientUser)-[crxn:REACTS_TO_POST]->(post)
			OPTIONAL MATCH (clientUser)-[csaves:SAVES_POST]->(post)
			OPTIONAL MATCH (clientUser)-[creposts:REPOSTS_POST]->(post)
	
			WITH post, 
				toString(post.created_at) AS created_at, 
				ownerUser { .username, .profile_pic_url } AS owner_user,
				CASE crxn 
					WHEN IS NULL THEN "" 
					ELSE crxn.reaction 
				END AS client_reaction, 
				CASE csaves 
					WHEN IS NULL THEN false 
					ELSE true 
				END AS client_saved, 
				CASE creposts 
					WHEN IS NULL THEN false 
					ELSE true 
				END AS client_reposted
			ORDER BY post.created_at DESC
			LIMIT 500
			RETURN collect(post { .*, owner_user, created_at, client_reaction, client_saved, client_reposted }) AS res_posts
		`
	} else {
		query = `
			MATCH (ownerUser:User)-[:CREATES_POST]->(post:Post { type: $type } WHERE post.description CONTAINS $term)
	
			WITH post, toString(post.created_at) AS created_at, ownerUser { .username, .profile_pic_url } AS owner_user
			ORDER BY post.created_at DESC
			LIMIT 500
			RETURN collect(post { .*, owner_user, created_at, client_reaction: "", client_saved: false, client_reposted: false }) AS res_posts
		`
	}

	res, err := db.Query(
		ctx,
		query,
		map[string]any{
			"type":            postType,
			"term":            term,
			"client_username": clientUsername,
		},
	)
	if err != nil {
		log.Println("contentRecommendationService.go: FetchPosts:", err)
		return nil, fiber.ErrInternalServerError
	}

	if len(res.Records) == 0 {
		return nil, nil
	}

	resPosts, _, _ := neo4j.GetRecordValue[[]any](res.Records[0], "res_posts")

	return resPosts, nil
}
